### PR TITLE
Update doc for query errors and add unit tests for JsonParserIterator

### PR DESCRIPTION
**Pull request by jihoonson**
### Description

A follow-up of https://github.com/apache/druid/pull/10746. This PR updates the query error doc and adds unit tests for `JsonParserIterator`. Integration tests will be added in another PR.

<hr>

This PR has:
- [x] been self-reviewed.
   - [ ] using the [concurrency checklist](https://github.com/apache/druid/blob/master/dev/code-review/concurrency.md) (Remove this item if the PR doesn't have any relation to concurrency.)
- [x] added documentation for new or modified features or behaviors.
- [ ] added Javadocs for most classes and all non-trivial methods. Linked related entities via Javadoc links.
- [ ] added or updated version, license, or notice information in [licenses.yaml](https://github.com/apache/druid/blob/master/licenses.yaml)
- [ ] added comments explaining the "why" and the intent of the code wherever would not be obvious for an unfamiliar reader.
- [x] added unit tests or modified existing tests to cover new code paths, ensuring the threshold for [code coverage](https://github.com/apache/druid/blob/master/dev/code-review/code-coverage.md) is met.
- [ ] added integration tests.
- [ ] been tested in a test Druid cluster.

### CODE REVIEW

File: processing/src/main/java/org/apache/druid/query/ResourceLimitExceededException.java [hunk 1]
```diff
@@ -35,9 +34,14 @@
 {
   public static final String ERROR_CODE = "Resource limit exceeded";
 
-  public ResourceLimitExceededException(String message, Object... arguments)
+  public ResourceLimitExceededException(String message)
```
Comment (clintropolis) [inline processing/src/main/java/org/apache/druid/query/ResourceLimitExceededException.java:37]: hmm, the built in format seemed useful and cleaner than having callers deal with it everywhere, and consistent with many of our other exceptions, why this change?

Comment (jihoonson) [reply processing/src/main/java/org/apache/druid/query/ResourceLimitExceededException.java]: I made a mistake in my previous PR that creates a `ResourceLimitExceededException` by calling `new ResourceLimitExceededException(errorCode, message, errorClass, host)` which actually calls `ResourceLimitExceededException(String message, Object... arguments)`. I didn't notice my mistake until I add a test for it because Intellij says you are cool. This change was to avoid such mistakes, but I agree it is useful. I added a static constructor that does the same.

**Review by clintropolis — APPROVED** [review]

**Review by clintropolis — APPROVED** [review]
